rpcs/wsFetcher.go

```diff
@@ -139,66 +98,48 @@ func (wsf *WsFetcher) NumPeers() int {
 // Close calls a delegate close fn passed in by the parent of this fetcher
 func (wsf *WsFetcher) Close() {
 	wsf.f.Close()
-	if wsf.closeFn != nil {
-		wsf.closeFn(wsf)
-	}
 }
 
 // a stub fetcherClient to satisfy the NetworkFetcher interface
 type wsFetcherClient struct {
-	listeners map[uint64]chan WsGetBlockOut
-	target    network.UnicastPeer
-	tag       protocol.Tag
+	target      network.UnicastPeer                    // the peer where we're going to send the request.
+	tag         protocol.Tag                           // the tag that is associated with the request/
+	service     *WsFetcherService                      // the fetcher service. This is where we perform the actual request and waiting for the response.
+	pendingCtxs map[context.Context]context.CancelFunc // a map of all the current pending contexts.
 
-	closed bool
+	closed bool // a flag indicating that the fetcher will not perform additional block retrivals.
 
-	mu deadlock.RWMutex
+	mu deadlock.Mutex
 }
 
-func (w *wsFetcherClient) getChForRound(r uint64) chan WsGetBlockOut {
+// GetBlockBytes implements FetcherClient
+func (w *wsFetcherClient) GetBlockBytes(ctx context.Context, r basics.Round) ([]byte, error) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 	if w.closed {
-		return nil
-	}
-	v, ok := w.listeners[r]
-	if !ok {
-		v = make(chan WsGetBlockOut, numBufferedInternalMsg)
-		w.listeners[r] = v
+		return nil, fmt.Errorf("wsFetcherClient(%d): shutdown", r)
 	}
-	return v
-}
+	childCtx, cancelFunc := context.WithCancel(ctx)
+	w.pendingCtxs[childCtx] = cancelFunc
+	w.mu.Unlock()
 
-// GetBlockBytes implements FetcherClient
-func (w *wsFetcherClient) GetBlockBytes(ctx context.Context, r basics.Round) ([]byte, error) {
-	listen := w.getChForRound(uint64(r))
-	if listen == nil {
-		return nil, fmt.Errorf("wsFetcherClient(%d): preconnection closed", r)
-	}
+	defer func() {
+		cancelFunc()
+		w.mu.Lock()
```

Comment:
Is this an unbalanced Lock() call?

Reply:
no. see `defer w.mu.Unlock()` above it.

Reply:
This results in valid execution but is awkward to read.  Not worth rewriting to make it less awkward - how about a comment in the defer func() pointing out there's a defer unlock on the stack already?

Reply:
sure. I'll add that!